src/plugins/plexil/blackboard_adapter.cpp

```diff
@@ -0,0 +1,646 @@
+		return;
+	}
+	std::string uid;
+	params[0].getValue(uid);
+
+	std::unique_lock<std::mutex> lock(ifs_read_mutex_);
+	if (ifs_read_.find(uid) == ifs_read_.end()) {
+		logger_->log_warn("PlexilBB", "Invalid lookup %s unsubscribe for "
+		                  "unknown interface %s", state.name().c_str(), uid.c_str());
+	} else {
+		//logger_->log_debug("PlexilBB", "Unsubscribe for %s", state.toString().c_str());
+		debugMsg("BlackboardAdapter:unsubscribe", "Unsubscribe for " << state.toString());
+		auto range = subscribed_states_.equal_range(uid);
+		for (auto i = range.first; i != range.second; ++i) {
+			if (i->second == state) {
+				subscribed_states_.erase(i);
+				break;
+			}
+		}
+		if (subscribed_states_.count(uid) == 0) {
+			bbil_remove_data_interface(ifs_read_[uid]);
+			blackboard_->update_listener(this, BlackBoard::BBIL_FLAG_DATA);
+		}
+	}
+}
+
+void
+BlackboardPlexilAdapter::bb_interface_data_changed(fawkes::Interface *interface) throw()
+{
+	//logger_->log_debug("PlexilBB", "Event for %s", interface->uid());
+	interface->read();
+	auto range = subscribed_states_.equal_range(interface->uid());
+	for (auto i = range.first; i != range.second; ++i) {
+		if (i->second.name() == "BB_changed") {
+			// we need to issue a read call above (otherwise we would handle the
+			// value change with the old value) but this will cause Interface::changed()
+			// to always return false. However, since this is the handler for the
+			// data changed event, we feel confident to return true as value
+			m_execInterface.handleValueChange(i->second, PLEXIL::Value(true));
+		} else {
+			PLEXIL::StateCacheEntry entry;
+			lookupNow(i->second, entry);
+			m_execInterface.handleValueChange(i->second, entry.cachedValue()->toValue());
+		}
+	}
+	m_execInterface.notifyOfExternalEvent();
+}
+
+
+/** Perform given command.
+ * @param cmd command to execute
+ */
+void
+BlackboardPlexilAdapter::executeCommand(PLEXIL::Command* cmd)
+{
+	std::string const &name = cmd->getName();
+
+	auto c = commands_.find(name);
+	if (c != commands_.end()) {
+		c->second(cmd);
+	} else {
+		warn("NavGraphAdapter:executeCommand: called for unknown"
+		     " command " << name);
+		m_execInterface.handleCommandAck(cmd, PLEXIL::COMMAND_FAILED);
+		m_execInterface.notifyOfExternalEvent();
+	}
+}
+
+void
+BlackboardPlexilAdapter::bb_open_for_reading(PLEXIL::Command* cmd)
+{
+	std::vector<PLEXIL::Value> const &args = cmd->getArgValues();
+	if (! verify_args(args, "BlackboardAdapter:bb_open_for_reading",
+	                  {{"type", PLEXIL::STRING_TYPE},
+	                   {"id", PLEXIL::STRING_TYPE}}))
+	{
+		m_execInterface.handleCommandAck(cmd, PLEXIL::COMMAND_FAILED);
+		m_execInterface.notifyOfExternalEvent();
+		return;
+	}
+
+	std::string if_type;
+	std::string if_id;
+	args[0].getValue(if_type);
+	args[1].getValue(if_id);
+
+	std::string uid{if_type + "::" + if_id};
+
+	//logger_->log_debug("PlexilBB", "Opening %s for reading", uid.c_str());
```

Comment:
These are useful debug points, if something breaks and we chase a bug. Hence, it makes sense to leave those in for potential future use.

Reply:
ahaa ... I thought these were replaced with debugMsg since they come in succession most of the time.
But in that case, wouldn't it make sense to have a debug flag or log level or something?
anyways, I'm sure there's good reasoning behind that as well, but that's just my 2 cents on the issue.

Reply:
The debugMsg are there for when you debug your plexil code, they can give you additional inside.

The commented out logger statements are for the C++ adapter code. They are not a config flag because they will be necessary only when you work on (as in extend or debug) the C++ code, in which case you can simply comment in the respective code. There is no reason to take the (ever so small) performance hit for a run-time flag in this specific case.